docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,461 @@
+        - [Filter](#filter)
+        - [Sort](#sort)
+        - [Select](#select)
+        - [Limit](#limit)
+        - [Uniq](#uniq)
+        - [Metrics](#metrics)
+            - [Series & Histogram Operators](#series--histogram-operators)
+            - [Range Vector Operations](#range-vector-operations)
+    - [API Changes](#api-changes)
+        - [Faceting](#faceting)
+    - [Glossary](#glossary)
+    - [Syntax Types](#syntax-types)
+
+<!-- /TOC -->
+
+## Related Documents
+
+- [LogQL Pseudo Label Extraction Design Doc](https://docs.google.com/document/d/1BrxnzqEfZnVVYUnjpZR-b9i5w9FNWRzzhyOq5ftCc-Q/edit#)
+- [Original LogQL Design Doc](https://docs.google.com/document/d/1DAzwoZF0yD4oFSFcUlRcGG0Ent2myAlFr_lZc6es3fs/edit#heading=h.6u7w5ql6hn4p)
+- [Current LogQL Documentation](https://github.com/grafana/loki/blob/master/docs/logql.md)
+
+## LogQL Parsers
+
+A parser operator can only be applied to __log__ streams (a set of entries attached to a given set of label pairs) and will transform (or extract) them into __field__ streams.
+
+As quick overview let's consider stream below:
+
+```json
+{
+  "stream": {
+   "job": "loki-dev/ingester",
+   "cluster": "us-central1",
+   "namespace" : "loki-dev",
+  },
+  "values": [
+    [
+      "1584114643856675277",
+      "level=info ts=2020-03-13T15:50:43.856555 caller=metrics.go"
+    ],
+    [
+      "1584114643856675278",
+      "level=warn ts=2020-03-13T15:50:43.856556 caller=metrics.go"
+    ]
+  ]
+}
+```
+
+As you can see the log is using the popular logfmt so we could use the `| logfmt` parsers to transform those entries into the fields entries like below:
+
+```json
+{
+  "stream": {
+   "job": "loki-dev/ingester",
+   "cluster": "us-central1",
+   "namespace" : "loki-dev",
+  },
+  "values": [
+    [
+      "1584114643856675277",
+      {
+          "level":"info",
+          "ts": "2020-03-13T15:50:43.856555",
+          "caller": "metrics.go"
+      }
+    ],
+   [
+      "1584114643856675278",
+      {
+          "level":"warn",
+          "ts": "2020-03-13T15:50:43.856556",
+          "caller": "metrics.go"
+      }
+    ]
+  ]
+}
+```
+
+Field names are declared by each parsers, some parser will allow implicit fields extraction for __brevity__ like the `json` and `logfmt` parsers.
+
+Since field name can conflicts with labels during aggregation we will support for each parser the ability to rename a field. However in case of conflict fields will take precedence over labels.
+
+> Implicit fields extraction could means trouble in term of performance. However in most cases we will be able to infer all used fields from the full query and so limit the extraction per line.
+
+To cover a wide range of structured and unstructured logs, we will start with those four parsers: `glob`, `logfmt`, `json` and `regexp`.
+
+### Glob
+
+The glob parser (as described in previous design) `{job="prod/nginx"} | glob "*<client_ip> - - [ * ] \"* *<path> "` could be used to extract fields from nginx logs.
```

Comment:
Yes it will be a special glob, I used glob as it's the closest language but you're right it's not exactly that.

See aws example:

```
FIELDS @message
| PARSE @message "* [*] *" as loggingTime, loggingType, loggingMessage
| FILTER loggingType IN ["ERROR", "INFO"]
| DISPLAY loggingMessage, loggingType = "ERROR" as isError
```